### PR TITLE
Optionally return stored data after storing

### PR DESCRIPTION
**Pull request by jakirkham**
Related to issue ( https://github.com/dask/dask/issues/2156 )

Adds an argument Dask Array's `store`, called `keep`. When set, `keep` will return the stored result(s). The behavior of `keep` differs a little depending on whether `compute` is set. If `compute` is `False`, Dask Arrays corresponding to the stored values are returned, which can be chained into other computations, persisted, etc. If `compute` is `True`, then it is like calling `compute` on Dask Arrays that would have been returned if `keep` was `False`.

### CODE REVIEW

**Comment by jakirkham** [comment]
This works fine based on the tests added thus far. That said, would be good to get some feedback on this approach and whether it seems reasonable for Dask. Happy to provide more context if needed.

**Comment by mrocklin** [comment]
Interesting.  I like the approach.  Just so that I'm clear the objective here is something like 
```
x = ... load data
y = ... transform data
y = y.store(..., keep=True)
z = ... transform data further 
```

In this case it seems like `keep=True` should almost imply `compute=False`

**Comment by jakirkham** [comment]
Correct. That is the idea. Have a few other thoughts about how this can be used. Caching being one of them. Though there is already value in being able to checkpoint this computation.

Yeah was thinking about the meaning of `compute` in the context of `keep` as well. We can probably tweak the default arguments (e.g. make them `None` initially and then fill them out depending on what is set).

Another option would be to make `compute=True, keep=True` mean that `persist` is run on the Dask Array(s). Thus storing the data while loading it across the cluster. Storage in this case acts more like a checkpoint.

Something else to consider is we could break this up into `insert_to_ooc` and `retrieve_from_ooc` functions.  In this implementation `insert_to_ooc` wouldn't actually return any data, but would be chained to a corresponding `retrieve_from_ooc` function that would. When providing `compute=True, keep=True`, `persist` could be called on the `store` portion. This would allow the `store` portion to occur asynchronously and tie any `load` steps to their corresponding `store` step's `Future`. If the data is too large to actually persist (without spilling to disk) and the intermediate result is actually useful, this ends up being a nice option. Expect there are some good opportunities to explore loading data and caching here as well.

**Comment by mrocklin** [comment]
> Another option would be to make compute=True, keep=True mean that persist is run on the Dask Array(s). Thus storing the data while loading it across the cluster. Storage in this case acts more like a checkpoint.

I'm against this.  People should just use persist in this case.

> Something else to consider is we could break this up into insert_to_ooc and retrieve_from_ooc functions. In this implementation insert_to_ooc wouldn't actually return any data, but would be chained to a corresponding retrieve_from_ooc function that would. When providing compute=True, keep=True, persist could be called on the store portion. This would allow the store portion to occur asynchronously and tie any load steps to their corresponding store step's Future. If the data is too large to actually persist (without spilling to disk) and the intermediate result is actually useful, this ends up being a nice option. Expect there are some good opportunities to explore loading data and caching here as well.

I like this.  It seems fairly clean.  I think we've recommended this workflow anecdotally to people in the past.

**Comment by jakirkham** [comment]
Agreed. I'll retool this for the last proposal. Expect the code to get a lot cleaner (fewer branches) as a consequence.

**Comment by jakirkham** [comment]
Ok, I think this is starting to look the way we want. Would be curious to hear what you think.

File: dask/array/core.py [hunk 1]
```diff
@@ -891,20 +896,58 @@ def store(sources, targets, lock=True, regions=None, compute=True, **kwargs):
...
+                            dict(sharedict.merge(
+                                (
+                                    each_store_key,
+                                    {each_store_key: store_dsk[each_store_key]}
+                                ),
+                                dsk,
+                                src.dask
+                            ))
+                        ).persist().dask[each_store_key]
+                    )
```
Comment (mrocklin) [inline dask/array/core.py]: `.persist` is blocking when using the single machine scheduler.  We definitely don't want to call it in a loop.

FWIW I find this statement difficult to parse.  That's subjective though.

File: dask/array/core.py [hunk 2]
```diff
@@ -2566,29 +2609,61 @@ def concatenate(seq, axis=0, allow_unknown_chunksizes=False):
...
+                lock.release()
+
+        return result
+
+    dsk = dict()
+    for store_key in out_dsk:
+        load_key = ('load-%s' % store_key[0].lstrip("store-"),) + store_key[1:]
+        dsk[load_key] = (load, store_key,) + out_dsk[store_key][3:-1]
 
     return dsk
```
Comment (mrocklin) [inline dask/array/core.py]: Why is this necessary on top of `from_array`?  They seem very similar

**Comment by mrocklin** [comment]
To be clear, current machinery doesn't implement the necessary bits to do this currently, but my hope was that we could modify the existing machinery in relatively minor ways rather than create a parallel system alongside.

**Comment by mrocklin** [comment]
There may be reasons against this of course.  As a disclaimer I haven't yet given this a very thorough review.

File: dask/array/core.py (hunk 1 above)
Comment (jakirkham) [reply dask/array/core.py]: Sure we can drop all of these into one `persist` call. Expect that will make it easier to parse as well.

File: dask/array/core.py (hunk 2 above)
Comment (jakirkham) [reply dask/array/core.py]: 😕 I don't think we are calling `from_array` or are you meaning that you feel the behavior is similar to `from_array`? If the latter, more on this in the comment to follow.

File: dask/array/core.py (hunk 1 above)
Comment (jakirkham) [reply dask/array/core.py]: Should add that generally what we are trying to accomplish here is submitting the Dask Graph associated with storing each chunk.

File: dask/array/core.py (hunk 2 above)
Comment (mrocklin) [reply dask/array/core.py]: This logic seemed similar to `from_array`.  I was curious if there was a reason why we can't reuse the existing functionality.  Mostly I'm concerned with the amount of code being written into dask.array.core to support this feature.  If it isn't easy to accomplish what we want to do with existing tools then is there something we can do to improve our current tools?

**Comment by jakirkham** [comment]
Thanks for the feedback.

From a big picture standpoint, we'd really like to be able to view results as they are being computed. Also we'd really like to make sure that what we view is an accurate reflection of what we have saved to disk. In order to handle both of these, we really need to be able to view a chunk as soon as its done computing and saved to disk even if other ones are still being computed and/or written. Hence having this one-to-one mapping of store to load steps aids that goal by letting us pull in computed chunks from disk once they are ready. It's pretty easy to handle the problem in this context as we have all the relevant pieces. Moving them elsewhere may be possible, but could make it harder to reason about.

Ultimately I'd like to condense the branching in this PR. Expect that the end result will be able to reuse the bulk of `store` for the various different use cases. However it has been easier to try and deal with this as a separate code path initially. Mainly have done this to avoid locking us into a particular way of handling this case initially until we are happy with what it is doing. Also have done this to avoid messing up special case handling at the end of `store` or dealing with some special cases that `store` has like using `delayed` objects for output. IOW this is a consequence of where I have spent my time initially and not how I expect it to look in finished form.

**Comment by mrocklin** [comment]
OK, I see how what I was proposing doesn't achieve the same goals as what you're doing here.  I agree that what you're proposing sounds attractive.

File: dask/array/core.py (hunk 1 above)
Comment (jakirkham) [reply dask/array/core.py]: Also valuable to combine `persist` calls to reuse intermediates.

FWIW I think this statement should now be a bit cleaner and more performant.

**Comment by jakirkham** [comment]
This could use another look if you have time.

**Comment by jakirkham** [comment]
Will probably do some more work on the tests, but I think this is ready for a more thorough review.

**Comment by mrocklin** [comment]
Thanks.  It is unlikely that I will get to this personally before next
Monday. I have a few personal things going on today an this weekend.

File: dask/array/core.py [hunk 3]
```diff
@@ -893,20 +897,61 @@ def store(sources, targets, lock=True, regions=None, compute=True, **kwargs):
...
 
-    name = 'store-' + tokenize(*keys)
-    dsk = sharedict.merge({name: keys}, updates)
-    result = Delayed(name, dsk)
-    if compute:
-        result.compute()
+    if keep:
+        if compute:
+            from ..base import persist
+            store_dlyds = persist(*store_dlyds)
```
Comment (mrocklin) [inline dask/array/core.py]: I think that we lose the dask.array optimizations here.  Being able to call `persist` on a dask.array here would be advantageous.  I probably gave bad advice in https://github.com/dask/dask/pull/3000 as well.  Will comment there shortly.

Comment (mrocklin) [reply dask/array/core.py]: If I recall correctly your previous solution involved a `map_blocks` call that assigned the value to the store and then returned that input downstream.  I apologize if I dissuaded you from this approach.  

Sorry in general for being short on time recently.

Comment (jakirkham) [reply dask/array/core.py]: No worries. It's the holidays after all. :)

Hmm...not sure that I follow the `persist` suggestion. While a Dask Array does go in as input, currently we are returning `out`, which could be...anything? Does `Delayed` not make sense here? How do we call `persist` on the `store` step without wrapping it in `Delayed`?

I don't think I was using `map_blocks` previously. Though I guess we could look at that option too. Is it ok for `map_block` to return an `object` array? Would it coerce `out` into something in memory (e.g. thinking of a persistent Zarr Array converted into a NumPy Array)?

Comment (mrocklin) [reply dask/array/core.py]: When we call compute/persist we first optimize the graph based on the type of the collection.  In the case of delayed objects there isn't much optimization to do.  However when the collection is a dask.array a ton of stuff gets run that fuses slicing, atop calls (eventually), and other special cases. By moving the task graph from an array to a delayed object without going through an optimization phase we've sidestepped calling optimizations on that graph that should have been called.

It is fine returning a delayed object, but we should call `__dask_optimize__` on the task graph before letting go of the array.

Comment (jakirkham) [reply dask/array/core.py]: Of course, that sounds totally reasonable. Happy to add a call to `__dask_optimize__` earlier on.

Comment (jakirkham) [reply dask/array/core.py]: Did this for `src` before calling `insert_to_ooc` in the `for`-loop. So will be reused for all the other cases. Not sure if that is what you had in mind though.

**Comment by jakirkham** [comment]
Any thoughts on this?

File: dask/array/core.py [hunk 4]
```diff
@@ -893,20 +897,72 @@ def store(sources, targets, lock=True, regions=None, compute=True, **kwargs):
...
+            src = Array(
+                src.__dask_optimize__(
+                    src.__dask_graph__(),
+                    src.__dask_keys__()
+                ),
+                src.name,
+                src.chunks,
+                src.dtype,
+                src.shape
+            )
```
Comment (mrocklin) [inline dask/array/core.py]: The previous `src` object here is a dask array?  If so then is there a reason why we call optimize before passing it to another array?  We should only need to call optimize when we leave the array collection type.

File: dask/array/core.py [hunk 5]
```diff
@@ -893,20 +897,72 @@ def store(sources, targets, lock=True, regions=None, compute=True, **kwargs):
...
+                    sharedict.merge(
+                        (
+                            each_store_key,
+                            {each_store_key: each_store_dsk[each_store_key]}
+                        ),
+                        dsk,
+                        src.dask
+                    )
+                )
+            )
```
Comment (mrocklin) [inline dask/array/core.py]: It would also be fine here (and possibly more efficient) to just make a single `dsk` object for all of the delayed objects and differentiate them only by their key.  It is ok to have excess keys in a task graph.

Comment (mrocklin) [reply dask/array/core.py]: I'm assuming that the number of elements in each_store_dsk is proportional to the number of chunks.

File: dask/array/core.py [hunk 6]
```diff
@@ -2585,15 +2645,41 @@ def store(x, out, index, lock, region):
...
+        finally:
+            if lock:
+                lock.release()
+
+        return result
+
+    dsk = dict()
+    for store_key in out_dsk:
+        load_key = ('load-%s' % store_key[0][len("store-"):],) + store_key[1:]
+        dsk[load_key] = (load, store_key,) + out_dsk[store_key][3:-1]
```
Comment (mrocklin) [inline dask/array/core.py]: If possible it would be nice to avoid placing dynamically generated functions (like `load` above) into task graphs.  This tends to be a bottleneck for larger computations on distributed machines where function serialization can become an issue.

File: dask/array/core.py [hunk 7]
```diff
@@ -2585,15 +2645,41 @@ def store(x, out, index, lock, region):
...
     dsk = dict()
     for t, slc in zip(core.flatten(arr.__dask_keys__()), slices):
         store_key = (name,) + t[1:]
-        dsk[store_key] = (store, t, out, slc, lock, region)
+        dsk[store_key] = (store, t, out, slc, lock, region, keep)
+
+    return dsk
+
+
+def retrieve_from_ooc(out_dsk):
```
Comment (mrocklin) [inline dask/array/core.py]: This could use a docstring

File: dask/array/core.py (hunk 4 above)
Comment (jakirkham) [reply dask/array/core.py]: Sorry not following. This was my understanding of your suggestion in the thread above.

There is actually only one case where we generate a Dask Array and I'm not sure that we shouldn't optimize it in that case too given we are trying to write this efficiently.

File: dask/array/core.py (hunk 5 above)
Comment (jakirkham) [reply dask/array/core.py]: For the most part, we are not using the `Delayed` objects as can be seen in the code below. We actually unpack them in most cases and constructing some other object.

The one exception is `compute=True, keep=True` where we `persist` them. In that case, we'd still like to keep separate `Delayed` objects for each chunk as we'd like to be able to get some results from these Dask Arrays without having to block for the full result.

File: dask/array/core.py (hunk 6 above)
Comment (jakirkham) [reply dask/array/core.py]: Hmm...so this is simply mirroring the behavior of `insert_to_ooc` with `store`. If having these functions dynamically generated is an issue, it seems like we should start by fixing `insert_to_ooc`.

File: dask/array/core.py (hunk 7 above)
Comment (jakirkham) [reply dask/array/core.py]: That's fine. Again this is mirroring `insert_to_ooc`, which didn't have one either. Should `insert_to_ooc` have a docstring? FWIW this is part of the private API and shouldn't be user facing.

File: dask/array/core.py (hunk 4 above)
Comment (mrocklin) [reply dask/array/core.py]: The danger is when we take a graph out of a dask.array and put it into a delayed object, or just compute it.  Then the array-specific optimizations don't run.  

There is relatively little danger when putting a task graph into a dask array, unless that task graph comes from some other collection with non-trivial optimizations (like bag or dataframe)

File: dask/array/core.py (hunk 5 above)
Comment (mrocklin) [reply dask/array/core.py]: We can still have separate delayed objects that share the same task graph.  Something like the following:

```python
dsk = {'a': (inc, 1), 'b': (dec, 1)}
a = Delayed('a', dsk)
b = Delayed('b', dsk)
```

In this case we constructed two delayed objects, but only constructed a single task graph.  I anticipate that generating millions of ShareDicts might be somewhat expensive (though we've tried hard to reduce these costs).  I would instead create a single task graph with all of the relevant tasks for all of the delayed objects and then just hand that single task graph to all of the delayed objects with distinct keys.

File: dask/array/core.py (hunk 6 above)
Comment (mrocklin) [reply dask/array/core.py]: `da.store` was left derelict for some time.  We just never touched it after the change to the dask.distributed scheduler.  My apologies for the stale code.

File: dask/array/core.py (hunk 7 above)
Comment (mrocklin) [reply dask/array/core.py]: It is handy when reviewing.  I had no idea what `out_dsk` represented for example.

File: dask/array/core.py (hunk 5 above)
Comment (jakirkham) [reply dask/array/core.py]: Ah ok. Sorry, don't think I was following what you were suggesting before. This make sense.

File: dask/array/core.py (hunk 4 above)
Comment (jakirkham) [reply dask/array/core.py]: Right. So the only case that I can think of that wouldn't need this optimization would be `compute=False, keep=True`. Would you like a guard against optimizing in that case? All the other cases create `Delayed` object(s).

Comment (mrocklin) [reply dask/array/core.py]: When we create those delayed objects we point them to graphs that have array computations in them.  Have those array tasks already gone through an optimization phase?

### This is bad

```python
x = da.from_array(...)
x[:10][:5]  # this will be optimized away

v = Delayed(key, x.dask)  # lose array optimizations at this point
```

### This is good

```python
x = da.from_array(...)
x[:10][:5]  # this will be optimized away

dsk = x.__dask_optimize__(x.__dask_graph__(), x.__dask_keys__())  # pre-optimize graph

v = Delayed(key, dsk) 
```

Comment (jakirkham) [reply dask/array/core.py]: Right. That's what this optimize call is doing.

Comment (mrocklin) [reply dask/array/core.py]: OK, I see, you're optimizing and then placing it back into the dask array.

What happens if the except branch is not triggered?  

Hrm, one issue that will come up when storing multiple arrays is that if there are multiple sources that have overlapping computations then optimizing them separately will likely create redundant work (dask arrays can fuse operations somewhat aggressively).  It looks like optimizing everything ahead of time might be a bit painful.

Comment (jakirkham) [reply dask/array/core.py]: Yeah, we still need a Dask Array object in `insert_to_ooc`. Though as we are just pulling out `keys` and finding corresponding `slice`s, I guess we could forego the Dask Array if we changed `insert_to_ooc`'s signature. Is there any performance penalty when stuffing this back into a Dask Array?

Are you referring to `except AttributeError` above? I think that is from a PR that allowed `Delayed` objects as targets. So my understanding is it was never meant to be anything more sophisticated than say a function that created a place to store the results. Honestly not sure it would even work with something more sophisticated than that even in its current form in `master`.

I see. So would it be best to get Dask graphs for everything before the `for`-loop and optimize them? Let me give that some thought. I think it should be possible.

Comment (mrocklin) [reply dask/array/core.py]: >  Is there any performance penalty when stuffing this back into a Dask Array?

I *think* that this should be more-or-less fine.  We should profile this on a many-chunked array before merging though and see if any new hot-spots arise.

**Comment by jakirkham** [comment]
Figured we should do a little cleanup on `insert_to_ooc` first. So made PR ( https://github.com/dask/dask/pull/3012 ) to do exactly this.

File: dask/array/core.py (hunk 4 above)
Comment (jakirkham) [reply dask/array/core.py]: > What happens if the except branch is not triggered?

Had an indentation error that I overlooked. 😞 Have since fixed it.

Comment (jakirkham) [reply dask/array/core.py]: > Hrm, one issue that will come up when storing multiple arrays is that if there are multiple sources that have overlapping computations then optimizing them separately will likely create redundant work (dask arrays can fuse operations somewhat aggressively). It looks like optimizing everything ahead of time might be a bit painful.

Tweaked the optimization to handle this case. Honestly not to bad since `__dask_optimize__` is a static method. Also the docs point out this can work on multiple collections. So this was basically designed to work. :)

File: dask/array/core.py (hunk 5 above)
Comment (jakirkham) [reply dask/array/core.py]: Have made some improvement on this by reusing the same task graph for each array being stored. As the task graph is unified and optimized over all `sources` earlier, this should pretty much address this issue.

**Comment by jakirkham** [comment]
So I think I have dealt with the optimization issues reasonably well. Would like to figure out how we would like to cleanup `insert_to_ooc`. Once that is dealt with we can cleanup `retrieve_from_ooc` accordingly.

**Comment by mrocklin** [comment]
@jcrist when you get back from vacation you might want to give this PR a look

File: dask/array/core.py [hunk 8]
```diff
@@ -2642,13 +2707,76 @@ def insert_to_ooc(arr, out, lock=True, region=None):
...
+    keys: Sequence
+        A sequence containing Dask graph keys to load
+    dsk: Mapping
+        A Dask graph corresponding to a Dask Array
+
+    Examples
+    --------
+    >>> import dask.array as da
+    >>> d = da.ones((5, 6), chunks=(2, 3))
+    >>> retrieve_from_ooc(d.dask.keys(), d.dask)  # doctest: +SKIP
```
Comment (jakirkham) [inline dask/array/core.py]: This is an inaccurate example. Will need to fix.

File: dask/array/core.py [hunk 9]
```diff
@@ -880,31 +884,85 @@ def store(sources, targets, lock=True, regions=None, compute=True, **kwargs):
...
+        each_store_dsk = insert_to_ooc(
+            src, tgt, lock=lock, region=reg, keep=keep
+        )
 
-    name = 'store-' + tokenize(*keys)
-    dsk = sharedict.merge((name, updates), *[src.dask for src in sources])
-    if compute:
-        compute_as_if_collection(Array, dsk, keys, **kwargs)
+        if keep:
+            load_names.append('load-store-%s' % src.name)
```
Comment (jakirkham) [inline dask/array/core.py]: Changed to `load-store-*` instead of `load-*`. It makes the code a little cleaner. Though if there is strong preference, this could be changed back.

**Comment by jakirkham** [comment]
Added one more test for `keep=True`. Seeing a bunch of failures related to `sparse` tests that I didn't touch. Are these already known issues?

**Comment by mrocklin** [comment]
Yeah, it looks like we're installing sparse from git-master.  A number of changes have taken place in the sparse codebase recently.  cc'ing @hameerabbasi in case he is interested in this.

**Comment by hameerabbasi** [comment]
I'll try to fix the tests and open a new pull request.

**Comment by mrocklin** [comment]
@hameerabbasi was kind enough to resolve the `sparse` issues.  It looks like there might be new pandas failures?

(Appveyor continued to fail with sparse issues.  I'm rerunning now)

**Comment by jakirkham** [comment]
Thanks for doing that. Raised the `pandas` failures as issue ( https://github.com/dask/dask/issues/3039 ).

**Comment by jakirkham** [comment]
So `pandas` test issues aside, where do we stand with this code? Is this ok? Are there other changes needed?

**Comment by mrocklin** [comment]
I'm fine with these changes.  I suspect that @jcrist might have thoughts,
but he may be out of contact while still on holidays.  If this is holding
you back from making progress I'm not against merging and reviewing
retroactively.  This PR has been up for a long while.

File: dask/array/core.py [hunk 10]
```diff
@@ -842,6 +842,8 @@ def store(sources, targets, lock=True, regions=None, compute=True, **kwargs):
         for the corresponding source and target in sources and targets, respectively.
     compute: boolean, optional
         If true compute immediately, return ``dask.delayed.Delayed`` otherwise
+    keep: boolean, optional
```
Comment (jcrist) [inline dask/array/core.py]: `keep` doesn't feel like the best keyword, but I struggle to think of a better one. `return_stored`?

File: dask/array/core.py [hunk 11]
```diff
@@ -880,31 +884,85 @@ def store(sources, targets, lock=True, regions=None, compute=True, **kwargs):
...
+            each_tgt_dsk = {}
 
-        update = insert_to_ooc(src, tgt, lock=lock, region=reg)
-        keys.extend(update)
+        src = Array(
+            sources_dsk,
+            src.name,
+            src.chunks,
+            src.dtype
+        )
```
Comment (jcrist) [inline dask/array/core.py]: Just a stylistic choice, I personally prefer to do all this one line. I find the ability to read isn't harmed, and the reduced number of lines makes the overall function easier to grasp. There are a few places in this PR where I'd change this. No strong thoughts though, up to you.

```python
        src = Array(sources_dsk, src.name, src.chunks, src.dtype)
```

File: dask/array/core.py [hunk 12]
```diff
@@ -880,31 +884,85 @@ def store(sources, targets, lock=True, regions=None, compute=True, **kwargs):
...
+
+        store_keys.extend(each_store_dsk.keys())
+        store_dsks.append(each_store_dsk)
+
+    store_dsks_mrg = sharedict.merge(*concatv(
+        store_dsks, tgt_dsks, [sources_dsk]
+    ))
+
+    result = None
+    if keep:
```
Comment (jcrist) [inline dask/array/core.py]: I'd personally change this logic to:

```python
if keep:
    if compute:
        ...
    return tuple(result)
elif compute:
    dyld.compute()
    return None
else:
    return dyld
```

Reduces nesting, and doesn't set `result` to `None` far from where it's used. Again, just a personal preference, no strong thoughts.

File: dask/array/core.py [hunk 13]
```diff
@@ -880,31 +884,85 @@ def store(sources, targets, lock=True, regions=None, compute=True, **kwargs):
...
+        if compute:
+            from ..base import persist
```
Comment (jcrist) [inline dask/array/core.py]: Should be able to move this import to the file level.

File: dask/array/core.py [hunk 14]
```diff
@@ -859,6 +861,8 @@ def store(sources, targets, lock=True, regions=None, compute=True, **kwargs):
 
     >>> store([x, y, z], [dset1, dset2, dset3])  # doctest: +SKIP
     """
+    from ..delayed import Delayed
```
Comment (jcrist) [inline dask/array/core.py]: Should be able to move this import to the file level.

File: dask/array/core.py [hunk 15]
```diff
@@ -2642,13 +2707,78 @@ def insert_to_ooc(arr, out, lock=True, region=None):
...
+    >>> d = da.ones((5, 6), chunks=(2, 3))
+    >>> a = np.empty(d.shape)
+    >>> g = insert_to_ooc(d, a)
+    >>> retrieve_from_ooc(g.keys(), g)  # doctest: +SKIP
+    """
+
+    load_dsk = dict()
+    for each_key in keys:
+        load_key = ('load-%s' % each_key[0],) + each_key[1:]
+        load_dsk[load_key] = (load_chunk, each_key,) + dsk[each_key][3:-1]
```
Comment (jcrist) [inline dask/array/core.py]: Slight preference to write as a comprehension:

```python
return {('load-%s' % k[0],) + k[1:]: (load_chunk, k,) + dsk[k][3:-1]
        for k in keys}
```

Comment (jcrist) [reply dask/array/core.py]: Also, can you explain the `[3:-1]` select here in a comment?

File: dask/array/core.py (hunk 10 above)
Comment (jakirkham) [reply dask/array/core.py]: Yeah I liked `keep` as it was short and seemed clear (but maybe not for everyone). Guess `return` is out. 😜 Could go with `return_stored`.

File: dask/array/core.py (hunk 14 above)
Comment (jakirkham) [reply dask/array/core.py]: I thought that too, but had seen that `store` was doing this import internally previously for the `compute=False` case. 😕 Not sure if there was a reason for it or if it still mattered.

File: dask/array/core.py (hunk 11 above)
Comment (jakirkham) [reply dask/array/core.py]: Yeah think I ran into issues with line length/`flake8` at some point. Can give it another look though as the code has changed over time.

File: dask/array/core.py (hunk 12 above)
Comment (jakirkham) [reply dask/array/core.py]: Ok, I'll give it a look.

File: dask/array/core.py (hunk 11 above)
Comment (jcrist) [reply dask/array/core.py]: In cases like those I usually prefer something like

```python
        src = Array(sources_dsk, src.name,
                    src.chunks, src.dtype)
```

to minimize number of lines. But again, all personal preference.

File: dask/array/core.py (hunk 13 above)
Comment (jakirkham) [reply dask/array/core.py]: Ok.

File: dask/array/core.py (hunk 15 above)
Comment (jakirkham) [reply dask/array/core.py]: IDK for some reason that feels less readable to me. However its just personal preference and if there are strong feelings can change.

Yeah this is a little weird. Basically we know we want to keep those argument from the call to `store` previously. Happy to add a comment for it.

Comment (jcrist) [reply dask/array/core.py]: I personally prefer comprehensions (as long as they're not too complicated), as they make it immediately clear that the loop is just building a dictionary.

A comment would be nice for future maintainers. Thanks.

File: dask/array/core.py (hunk 10 above)
Comment (jcrist) [reply dask/array/core.py]: I find `return_stored` clearer, but there may be a better option.

File: dask/array/core.py (hunk 14 above)
Comment (jakirkham) [reply dask/array/core.py]: Should add there are a bunch of similar imports littered throughout the code.

Comment (jcrist) [reply dask/array/core.py]: Sometimes local imports are necessary. This also isn't a huge issue, just feels like if you're touching this code it's worth cleaning it up here at this time.

File: dask/array/core.py (hunk 12 above)
Comment (jakirkham) [reply dask/array/core.py]: Maybe I'm not understanding the concern here. The two `compute` cases now share a fair bit of code. So breaking them apart will create some redundancy.

Is the concern more that the `return` statement has an unclear value? If so, we could add `return` statements to the respective branches instead. Would that help?

File: dask/array/core.py (hunk 10 above)
Comment (jakirkham) [reply dask/array/core.py]: Ok, done.

File: dask/array/core.py (hunk 12 above)
Comment (jcrist) [reply dask/array/core.py]: > Is the concern more that the return statement has an unclear value? If so, we could add return statements to the respective branches instead. Would that help?

That's what I was suggesting, the dedent of the non-keep logic required a bit of duplication which I hadn't thought about. To be perfectly clear, here's my suggestion:

```python
if keep:
    if compute:
	store_dlyds = [Delayed(k, store_dsks_mrg) for k in store_keys]
	store_dlyds = persist(*store_dlyds)
	store_dsks_mrg = sharedict.merge(*[e.dask for e in store_dlyds])

    load_dsks_mrg = sharedict.merge(store_dsks_mrg, *load_dsks)

    result = []
    for each_load_name in load_names:
	result.append(Array(
	    load_dsks_mrg,
	    each_load_name,
	    src.chunks,
	    src.dtype
	))
    return tuple(result)
else:
    name = 'store-' + tokenize(*store_keys)
    dsk = sharedict.merge({name: store_keys}, store_dsks_mrg)
    dlyd = Delayed(name, dsk)

    if compute:
	dlyd.compute()
        return None
    else:
	return dlyd
```

Again, not a huge deal, just a suggestion.

File: dask/array/core.py (hunk 14 above)
Comment (jakirkham) [reply dask/array/core.py]: Agreed. Just wanted to make sure I wasn't missing anything before doing it. Have made this change.

File: dask/array/core.py (hunk 11 above)
Comment (jakirkham) [reply dask/array/core.py]: Seems this now fits on one line. Probably was spread out due to older less clear code. Now on one line.

File: dask/array/core.py (hunk 13 above)
Comment (jakirkham) [reply dask/array/core.py]: Done.

File: dask/array/core.py (hunk 15 above)
Comment (jakirkham) [reply dask/array/core.py]: Have left the `for`-loop as is for now.

Added a comment to explain what is happening here. Please let me know your thoughts on it.

**Comment by jakirkham** [comment]
Thanks @jcrist. Pushed a bunch of changes based on your suggestions above. There is one point regarding the `return` statement that I'm not quite following. Will happily correct that once I have a clearer picture of what you are looking for. Please let me know your thoughts on the other changes.

File: dask/array/core.py [hunk 16]
```diff
@@ -2642,13 +2702,84 @@ def insert_to_ooc(arr, out, lock=True, region=None):
...
+
+    load_dsk = dict()
+    for each_key in keys:
+        load_key = ('load-%s' % each_key[0],) + each_key[1:]
+        # Reuse the arguments from `store_chunk` in `load_chunk`. Knowing that
+        # `dsk[each_key]` has the format below, we need to make a small change.
+        # `(store_chunk, t, out, slc, lock, region, return_stored)`
+        # Namely drop the first 3 arguments as they are the function and two
+        # arrays that were already used up by `store_chunk`. In their place use
+        # `load_chunk` and the result from `store_chunk` (i.e. `each_key`).
```
Comment (jcrist) [inline dask/array/core.py]: Verbosity makes it harder for people to read when skimming code. I'd just write:

```python
# Reuse extra arguments from `store_chunk` in `load_chunk`.
```

Again, not a huge deal. I don't mean to bikeshed your code, I think we have two very different styles when writing. The comment is fine as is.

File: dask/array/core.py (hunk 12 above)
Comment (jakirkham) [reply dask/array/core.py]: Thanks. That's clear. Have updated.

File: dask/array/core.py (hunk 16 above)
Comment (jakirkham) [reply dask/array/core.py]: No worries. Wasn't sure the level of detail we wanted here. Happy to cut to one line.

Comment (jakirkham) [reply dask/array/core.py]: Done.

**Comment by jakirkham** [comment]
Think that addresses the last round of comments. Please let me know if there is anything else.

File: dask/array/core.py [hunk 17]
```diff
@@ -880,31 +885,79 @@ def store(sources, targets, lock=True, regions=None, compute=True, **kwargs):
...
+
+        result = []
+        for each_load_name in load_names:
+            result.append(Array(
+                load_dsks_mrg,
+                each_load_name,
+                src.chunks,
+                src.dtype
+            ))
+        result = tuple(result)
```
Comment (jcrist) [inline dask/array/core.py]: Apologies. One last nit: this loop/append/tuple can be reduced to a one liner, which IMO improves readability.

```python
        return tuple(Array(load_dsks_mrg, name, src.chunks, src.dtype)
                     for name in load_names)
```

Take it or leave it. Otherwise this looks good to me.

File: dask/array/core.py [hunk 18]
```diff
@@ -880,31 +885,79 @@ def store(sources, targets, lock=True, regions=None, compute=True, **kwargs):
...
                          % (len(sources), len(targets), len(regions)))
 
-    updates = {}
-    keys = []
+    # Optimize all sources together
+    sources_dsk = sharedict.merge(*[e.__dask_graph__() for e in sources])
+    sources_dsk = Array.__dask_optimize__(
+        sources_dsk,
+        [e.__dask_keys__() for e in sources]
+    )
```
Comment (jakirkham) [inline dask/array/core.py]: While you are looking @jcrist, would be good if you could give these optimization lines a quick look. Based on the docs this seemed ok, but you certainly know better since you wrote all of these `__dask_*__` functions. 😉

Comment (jcrist) [reply dask/array/core.py]: That's exactly how they should be used.

Comment (jakirkham) [reply dask/array/core.py]: Thanks for checking.

File: dask/array/core.py (hunk 17 above)
Comment (jakirkham) [reply dask/array/core.py]: Sure. Done.

File: dask/array/core.py (hunk 15 above)
Comment (jakirkham) [reply dask/array/core.py]: Just for extra clarity, `dsk[k][3:-1]` translates to `(index, lock, region)` for that `k`.

File: dask/array/core.py [hunk 19]
```diff
@@ -2584,15 +2629,21 @@ def store_chunk(x, out, index, lock, region):
...
 
     >>> a = np.ones((5, 6))
     >>> b = np.empty(a.shape)
-    >>> store_chunk(a, b, (slice(None), slice(None)), False, None)
+    >>> store_chunk(a, b, (slice(None), slice(None)), False, None, False)
     """
 
+    result = None
+    if return_stored:
+        result = out
```
Comment (jcrist) [inline dask/array/core.py]: Sorry, I missed this bit before. Is this still used? From below it looks like the result is always re-read from the store, but I may be missing something?

Comment (jakirkham) [reply dask/array/core.py]: Yes, this is still used.

In an effort to chain the `store_chunk` calls to the `load_chunk` calls, we return `out` unchanged so that it can be fed as an argument into `load_chunk` via a key. Does that make sense? It could be there is a better way to do this. Just seemed simple enough to implement at the time.

We could also move this branch to the end and simply `return` either `out` or `None` if that would make it more clear.

Comment (jcrist) [reply dask/array/core.py]: Ah, missed that it was `out` and not `x`. Fine by me.

**Comment by jcrist** [comment]
Thanks @jakirkham!

**Comment by jakirkham** [comment]
Thanks @mrocklin and @jcrist for the reviews.

**Comment by mrocklin** [comment]
Thank you @jakirkham for both your work and your patience here.

File: dask/array/core.py (hunk 17 above)
Comment (jakirkham) [reply dask/array/core.py]: Hmm...I think we got too clever here and introduced a bug. Shouldn't be `src` as that would only repeat the last array's metadata for all arrays. Will submit a follow-up PR to fix this.

Comment (jakirkham) [reply dask/array/core.py]: Added a test to demonstrate the bug and included a fix for it in PR ( https://github.com/dask/dask/pull/3064 ).

File: dask/array/core.py [hunk 20]
```diff
@@ -880,31 +885,73 @@ def store(sources, targets, lock=True, regions=None, compute=True, **kwargs):
...
     else:
-        from ..delayed import Delayed
-        dsk.update({name: keys})
-        return Delayed(name, dsk)
+        name = 'store-' + tokenize(*store_keys)
+        dsk = sharedict.merge({name: store_keys}, store_dsks_mrg)
+        result = Delayed(name, dsk)
+
+        if compute:
+            result.compute()
```
Comment (jakirkham) [inline dask/array/core.py]: Missed that this call should still get `**kwargs`. Fixing with PR ( https://github.com/dask/dask/pull/3300 ).
